src/main/modules/oidc/index.ts

```diff
@@ -97,11 +83,63 @@ export class OidcMiddleware {
     });
   }
 
-  private saveConfiguredSession(newSession: OIDCSession, session: Partial<AppSession>): Promise<AppSession> {
-    return new Promise((resolve) => {
-      session.user = newSession.user;
-      session.tokens = newSession.tokens;
-      session.save(() => resolve(session as AppSession));
+  private getSessionStore(app: Application): any {
+    const redisStore = ConnectRedis(session);
+    const fileStore = FileStoreFactory(session);
+
+    const redisHost: string = config.get('session.redis.host');
+    const redisPort: number = config.get('session.redis.port');
+    const redisPass: string = config.get('session.redis.key');
+
+    if (redisHost && redisPass) {
+      const client = createClient({
+        host: redisHost,
+        password: redisPass,
+        port: redisPort ?? 6380,
+        tls: true
+      });
+
+      app.locals.redisClient = client;
+      return new redisStore({ client });
+    }
+
+    return new fileStore({ path: '/tmp' });
+  }
+
+  private cacheSystemAccount = (app: Application): void => {
+    let delay = 10 * 60;
+
+    this.getSystemUserAccessToken()
+      .then(tokenSet => {
+        app.locals.container.register({
+          systemAxios: asValue(this.createAuthedAxiosInstance(tokenSet.access_token))
+        });
+
+        delay = tokenSet.expires_in/2;
+        this.logger.info('Refreshed system user token. Refreshing again in: ' + Math.floor(delay/60) + 'mins');
+      })
+      .catch(() => this.logger.info('Failed to refresh system user token. Refreshing again in: ' + delay/60 + 'mins'))
+      .finally(() => setTimeout(this.cacheSystemAccount, delay * 1000, app));
+  }
+
+  private createAuthedAxiosInstance(accessToken: string): AxiosInstance {
+    return axios.create({
+      baseURL: config.get('services.idam.url.api'),
+      headers: {Authorization: 'Bearer ' + accessToken}
     });
   }
+
+  private async getSystemUserAccessToken(): Promise<TokenSet> {
```

Comment:
since we use the system user access token for somethings, and the users token for others, do we log when we perform an action on behalf of the user

Reply:
@nikola-naydenov-hmcts currently the only use of the system user access token is for authorising an axios instance that calls to the 'getAllRoles' endpoint. I don't think we have any logging related to this (apart from when theirs an error) as all it does is get the role data so it can build an internal map of role hierarchies.